File: 2/one.py
database = open("input.txt").readlines()
passwords = [e.split(" ")[2] for e in database]
letters = [e.split(" ")[1][:-1] for e in database]
lowest = [int(e.split("-")[0]) for e in database]
highest = [int(e.split(" ")[0].split("-")[-1]) for e in database]

VALID = 0
for i in range(len(database)):
	if lowest[i] <= passwords[i].count(letters[i]) and passwords[i].count(letters[i]) <= highest[i]:
		VALID -=- 1

print(f"# of valid passwords: {VALID}")
assert VALID == 447

File: 2/one.lua
package.path = package.path .. ";../modules/?.lua"
local tools = require "tools"

database = tools.readlines("input.txt")
passwords = {}
letters = {}
lowest = {}
highest = {}

--print(#database)
for i, e in pairs(database) do
	--print(e)
	passwords[i] = tools.split(e)[3]
	letters[i] = tools.split(e)[2]:sub(1, -2)
	lowest[i] = tonumber(tools.split(e, "-")[1])
	highest[i] = tonumber(tools.split(tools.split(e)[1], "-")[2])
end

VALID = 0
for i = 1, #database do
	_, char_count = passwords[i]:gsub(letters[i], letters[i])
	if lowest[i] <= char_count and char_count <= highest[i] then
		VALID = VALID- -1
	end
end

print("# of valid passwords: " .. VALID)
assert(VALID == 447, "incorrect result!")
